deep4cast/custom_layers.py

```diff
@@ -1,9 +1,91 @@
 # -*- coding: utf-8 -*-
 """Custom layers module."""
 
-from keras.layers import Layer, Dropout
+from keras.layers import Layer, Dropout, Dense, Lambda, Wrapper
 from keras import backend as K
 from keras.legacy import interfaces
+from keras import initializers
+from keras.engine import InputSpec
+
+
+class ConcreteDropout(Wrapper):
+    """Applies Dropout to one input layer, even at prediction time.
+    Dropout consists in randomly setting
+    a fraction `rate` of input units to 0 at each update during training time,
+    which helps prevent overfitting. Concrete Dropout provides a principled method
+    to learn an optimal dropout rate, jointly within the DNN training.
+    # Arguments
+        layer: an arbitrary Keras layer instance, e.g. Dense(), Conv2D(). 
+            The current implementation supports 2D inputs only for Conv layer.
+        weight_regularizer: a positive float number to regulate DNN weights.
+        dropout_regularizer: a positive float number to regulate the dropout.
+        seed: A Python integer to use as random seed.
+    # References
+        - [Concrete Dropout](https://papers.nips.cc/paper/6949-concrete-dropout.pdf)
+    """
+
+    def __init__(self, layer, weight_regularizer=1e-6, dropout_regularizer=1e-5, **kwargs):
+        # assert ’kernel_regularizer’ not in kwargs
+        super(ConcreteDropout, self).__init__(layer, **kwargs)
+        self.weight_regularizer = K.cast_to_floatx(weight_regularizer)
+        self.dropout_regularizer = K.cast_to_floatx(dropout_regularizer)
+        self.mc_test_time = mc_test_time
+        self.losses = []
+        self.supports_masking = True
+
+    def build(self, input_shape=None):
+        assert len(input_shape) == 2  # TODO: test with more than two dims
+        self.input_spec = InputSpec(shape=input_shape)
+        if not self.layer.built:
+            self.layer.build(input_shape)
+            self.layer.built = True
+        # this is very weird, we must call super before we add new losses
+        super(ConcreteDropout, self).build(input_shape)
+
+        # initialise p
```

Comment:
Could you provide some more commenting as to what each section of the code does? Also many functions don't have a doc string or commenting as to what they are supposed to be doing.